cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -173,22 +171,18 @@ private void addNotification(Job job, Notification notification) {
     put(newJob);
   }
 
-  private boolean isTriggerSatisfied(Trigger trigger, List<Notification> notifications) {
-    if (trigger instanceof TimeTrigger || trigger instanceof StreamSizeTrigger) {
-      // TimeTrigger/StreamSizeTrigger is satisfied as soon as the Notification arrive, due to how the Notification
-      // is initially created
-      return true;
+  private boolean isTriggerSatisfied(Trigger trigger, Notification notification) {
+    if (!(trigger instanceof SatisfiableTrigger)) {
+      // this shouldn't happen, since implementation of Trigger in ProgramSchedule should implement SatisfiableTrigger
+      throw new IllegalArgumentException("Implementation of Trigger in ProgramSchedule" +
+                                           " must implement SatisfiableTrigger");
     }
-    if (trigger instanceof PartitionTrigger) {
-      PartitionTrigger partitionTrigger = (PartitionTrigger) trigger;
-      int numPartitions = 0;
-      for (Notification notification : notifications) {
-        String numPartitionsString = notification.getProperties().get("numPartitions");
-        numPartitions += Integer.parseInt(numPartitionsString);
-      }
-      return numPartitions >= partitionTrigger.getNumPartitions();
+    if (((ProtoTrigger) trigger).getType().isImmediatelySatisfied()) {
```

Comment:
This is for backward compatibility. `updateStatus(notification)` for `TimeTrigger` and `StreamTrigger` look for fields in notifications that didn't exist in pre-4.3 versions. Since there can be unprocessed notifications in TMS from pre-4.3 version, calling `updateStatus(notification)` will return false even if the trigger is satisfied.